src/languages/gcode.js

```diff
@@ -7,58 +7,166 @@
  */
 
 export default function(hljs) {
-  const GCODE_IDENT_RE = '[A-Z_][A-Z0-9_.]*';
-  const GCODE_CLOSE_RE = '%';
+  const regex = hljs.regex;
   const GCODE_KEYWORDS = {
-    $pattern: GCODE_IDENT_RE,
-    keyword: 'IF DO WHILE ENDWHILE CALL ENDIF SUB ENDSUB GOTO REPEAT ENDREPEAT '
-      + 'EQ LT GT NE GE LE OR XOR'
-  };
-  const GCODE_START = {
-    className: 'meta',
-    begin: '([O])([0-9]+)'
+    $pattern: /[A-Z]+|%/,
+    keyword: [
+      // conditions
+      'THEN',
+      'ELSE',
+      'ENDIF',
+      'IF',
+
+      // controls
+      'GOTO',
+      'DO',
+      'WHILE',
+      'WH',
+      'END',
+      'CALL',
+
+      // scoping
+      'SUB',
+      'ENDSUB',
+
+      // comparisons
+      'EQ',
+      'NE',
+      'LT',
+      'GT',
+      'LE',
+      'GE',
+      'AND',
+      'OR',
+      'XOR',
+
+      // start/end of program
+      '%'
+    ],
+    built_in: [
+      'ATAN',
+      'ABS',
+      'ACOS',
+      'ASIN',
+      'COS',
+      'EXP',
+      'FIX',
+      'FUP',
+      'ROUND',
+      'LN',
+      'SIN',
+      'SQRT',
+      'TAN',
+      'EXISTS'
+    ]
   };
-  const NUMBER = hljs.inherit(hljs.C_NUMBER_MODE, { begin: '([-+]?((\\.\\d+)|(\\d+)(\\.\\d*)?))|' + hljs.C_NUMBER_RE });
+
+
+  // TODO: post v12 lets use look-behind, until then \b and a callback filter will be used
+  // const LETTER_BOUNDARY_RE = /(?<![A-Z])/;
+  const LETTER_BOUNDARY_RE = /\b/;
+
+  function LETTER_BOUNDARY_CALLBACK(matchdata, response) {
+    if (matchdata.index === 0) {
+      return;
+    }
+
+    const charBeforeMatch = matchdata.input[matchdata.index - 1];
+    if (charBeforeMatch >= '0' && charBeforeMatch <= '9') {
+      return;
+    }
+
+    response.ignoreMatch();
```

Comment:
If the regex is `![A-Z]` why wouldn't the conditional match?  ie:

```suggestion
    const charBeforeMatch = matchdata.input[matchdata.index - 1];
    if (charBeforeMatch >= 'A' && charBeforeMatch <= 'Z') {
      response.ignoreMatch();
    }
```

Reply:
Do we need to deal with lowercase a-z also?

Reply:
If the `begin' with `on:begin` variant was the only matching logic, we'd have to check for `![a-zA-Z]`.

But since we have a variant with `\b` we can be sure that any instance that follows a word boundary `![a-zA-Z0-9_]` has been found already. So we only need to additionally find those that follow `0-9` and ignore everything else.

Doing that means we don't have to account for the case.

I just noticed the underscore `_` isn't handled by that callback, but it should be. I will push another commit (but will wait for feedback on the thing above).

Reply:
Update: I've committed an additional check for underscores, to align the filter with what `\b` and `![A-Z]` mean.
I've added two lines to the markup test and verified in my v1 branch `feat/gcode-v12` that the result is the same.